### PR TITLE
wgengine/magicsock: send, use endpoints in CallMeMaybe messages

### LINKED ISSUE #1172: Exchange endpoints via DERP
Tracking bug for exchange endpoints via DERP, instead of only via control.

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes #1172

### CODE REVIEW

File: disco/disco.go [hunk 1]
```diff
@@ -124,13 +124,21 @@ func parsePing(ver uint8, p []byte) (m *Ping, err error) {
...
-	// server sends. But because the client could've been idle
-	// before it reached out to us, the control plane might
-	// have stale info and these endpoints in CallMeMaybe
-	// might contain the just-obtained-milliseconds-ago
-	// STUN endpoint.
+	//
+	// Prior to Tailscale 1.4, the endpoints were exchanged purely
+	// between nodes and the control server.
+	//
+	// Start with Tailscale 1.4, clients advertise their endpoints.
```
Comment (josharian) [inline disco/disco.go:131]: Starting

File: disco/disco.go [hunk 2]
```diff
@@ -124,13 +124,21 @@ func parsePing(ver uint8, p []byte) (m *Ping, err error) {
...
+	// Older clients won't use this, but newer clients should
+	// use any endpoints in here that aren't included from control.
+	//
+	// Control might have sent stale endpoints if the client was idle
+	// before contacting us. In that case, the client likely did a STUN
+	// request immediately before sending the CallMeMaybe to recreate
+	// their NAT port mapping, and that new good endpoint is included
+	// in this field, but might not yet be in control's endpoints.
+	// (And in the future, control will stop distributing endpoints
```
Comment (josharian) [inline disco/disco.go:140]: It occurs to me now (I'm slow) that in addition to being good for privacy, this also reduces the potential damage from a bad control release. Hurrah!

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -1882,6 +1882,27 @@ func (c *Conn) handlePingLocked(dm *disco.Ping, de *discoEndpoint, src netaddr.I
 	}, discoVerboseLog)
 }
 
+// enqueueCallMeMaybe schedules a send of disco.CallMeMaybe to de via derpAddr
+// once we know that our STUN endpoint is fresh.
+//
+// derpAddr is de.derpAddr at the time of send. It's assumed the peer won't be
+// flipping primary DERPs in the 0-30ms it takes to confirm our STUN endpoint.
```
Comment (josharian) [inline wgengine/magicsock/magicsock.go:1889]: If this does happen, what goes wrong? How long does it take to recover?

File: wgengine/magicsock/magicsock.go [hunk 4]
```diff
@@ -1882,6 +1882,27 @@ func (c *Conn) handlePingLocked(dm *disco.Ping, de *discoEndpoint, src netaddr.I
...
+func (c *Conn) enqueueCallMeMaybe(derpAddr netaddr.IPPort, de *discoEndpoint) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// TODO(bradfitz): do a fast/lite re-STUN if our STUN info is too old
+	// If it's too old, enqueue.
```
Comment (josharian) [inline wgengine/magicsock/magicsock.go:1895]: Can't parse these two comments taken together.

**Review by josharian — APPROVED** [review]
LGTM but might want to wait for Dave too

File: wgengine/magicsock/magicsock.go (hunk 3 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:1889]: clarified

File: wgengine/magicsock/magicsock.go (hunk 4 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:1895]: clarified

**Review by danderson — APPROVED** [review]

**Pull request merged**